src/lib/ipc/shmem/shmem.lua

```diff
@@ -0,0 +1,267 @@
+-- The shmem base class provides a simple IPC mechanism to exchange
+-- arbitrary cdata objects with other processes through a file-backed
+-- shared memeory region.
+--
+-- A new empty region is created by calling the constructor with a
+-- filename and optional directory, where the file should be created.
+--
+--  local shmem = require("lib.ipc.shmem.shmem")
+--  local foo = shmem:new({ filename = "foo", directory = "/tmp" })
+--
+-- If omitted, the directory defaults to "/tmp/snabb-shmem".  The
+-- constructor creates an empty file called the "data file" with the
+-- given name and maps it into the processe's virtual memory.  In
+-- addition, it creates an "index file" by appending the suffix
+-- ".index" to the data file name.  The first line of the index file
+-- contains a string that identifies the name space to which the
+-- memory region belongs, followed by a colon, followed by an integer
+-- version number.  The name space indicates how the rest of the index
+-- file needs to be interpreted.  A name space is tied to the subclass
+-- that implements it (but a subclass may inherit the name space of
+-- its ancestor class).  The version number allows for changes of the
+-- index format within the name space.
+--
+-- The base class provides the name space "default", i.e. the header
+-- line of the index file for version 1 contains the string "default:1".
+--
+-- The default name space contains a single line in the index file for
+-- every object stored in the memory region.  The order of the
+-- descriptions must be the same as that of the objects.  A
+-- description consists of an arbitrary name, followd by a colon,
+-- followed by the length of the corresponding object in bytes.  For
+-- example, the index file
+--
+--   default:1
+--   foo:4
+--   bar:7
+--
+-- Describes a memory region of length 11, which contains an object
+-- named 'foo' that consists of 4 bytes starting at offset 0 and an
+-- object named 'bar' consisting of 7 bytes starting at offset 4.  The
+-- type of the object is implied by its name and is not part of the
+-- description in the index. Each name must be unique.
+--
+-- An object is added to the region by calling the method register(),
+-- which takes a string, a ctype object and an optional value as
+-- arguments.  The ctype must refer to a complete type such that the
+-- size of the object is fixed when it is added to the index.  The
+-- following example adds an unsigned 32-bit number called "counter"
+-- and a struct named "bar":
+--
+--  local counter_t = ffi.typeof("uint32_t")
+--  local bar_t = ffi.typeof("struct { uint8_t x; char string[10]; }")
```

Comment:
I think the easiest would be some kind of double-buffering. Either with an atomic value to indicate which copy is the current one, or by having a monotonically increasing counter on the struct itself and just using the largest one.

It might be necessary some reordering barrier to make sure the critical atomic flag (the current index or the generation counter) is seen as updated last by every party.  Does anybody know what kind of non-reordering guarantees the x86 model provides?  I don't think `CAS` instructions apply too well on `mmap()`'ed blocks.